src/Mod/PartDesign/App/FeatureChamfer.cpp

```diff
@@ -138,16 +140,36 @@ App::DocumentObjectExecReturn *Chamfer::execute(void)
     try {
         BRepFilletAPI_MakeChamfer mkChamfer(baseShape.getShape());
 
-        TopTools_IndexedMapOfShape mapOfEdges;
         TopTools_IndexedDataMapOfShapeListOfShape mapEdgeFace;
         TopExp::MapShapesAndAncestors(baseShape.getShape(), TopAbs_EDGE, TopAbs_FACE, mapEdgeFace);
-        TopExp::MapShapes(baseShape.getShape(), TopAbs_EDGE, mapOfEdges);
 
-        for (std::vector<std::string>::const_iterator it=SubNames.begin(); it != SubNames.end(); ++it) {
-            TopoDS_Edge edge = TopoDS::Edge(baseShape.getSubShape(it->c_str()));
-            const TopoDS_Face& face = (chamferType != 0 && flipDirection) ?
-                TopoDS::Face(mapEdgeFace.FindFromKey(edge).Last()) :
-                TopoDS::Face(mapEdgeFace.FindFromKey(edge).First());
+        for (auto &itSN : SubNames) {
+            TopoDS_Edge edge = TopoDS::Edge(baseShape.getSubShape(itSN.c_str()));
+
+            const TopoDS_Shape& faceLast = mapEdgeFace.FindFromKey(edge).Last();
+            const TopoDS_Shape& faceFirst = mapEdgeFace.FindFromKey(edge).First();
+
+            // Set the face based on flipDirection for all edges by default. Note for chamferType==0 it does not matter which face is used.
+            TopoDS_Face face = TopoDS::Face( flipDirection ? faceLast : faceFirst );
+    
+            // for chamfer types otherthan Equal (type = 0) check if one of the faces associated with the edge
+            // is one of the originally selected faces. If so use the other face by default or the selected face if "flipDirection" is set
+            if (chamferType != 0) {
+
+                // for each selected face
+                for (auto &itFN : FaceNames) {
+                    const TopoDS_Shape selFace = baseShape.getSubShape(itFN.c_str());
+
+                    if ( faceLast.IsEqual(selFace) ) {
+                        face = TopoDS::Face( flipDirection ? faceFirst : faceLast );
+                    }
+                    else if ( faceFirst.IsEqual(selFace) ) {
+                        face = TopoDS::Face( flipDirection ? faceLast : faceFirst );
+                    }

```

Comment:
Totally inconsequential, but we typically don't use the curly braces on single-line conditionals (and that appears to be the usage in the rest of this file). So perhaps:
```suggestion
                    if ( faceLast.IsEqual(selFace) )
                        face = TopoDS::Face( flipDirection ? faceFirst : faceLast );
                    else if ( faceFirst.IsEqual(selFace) )
                        face = TopoDS::Face( flipDirection ? faceLast : faceFirst );
```

Reply:
@chennes 
Thanks for the feedback. I might be showing my age by saying this but the last time I did any serious C++ other than maintenance was back in the late 90's so C++98 level. Looks like need to do a refresher!

Happy to make the changes that you suggested, but given my rather embarrassing experience with the last PR I wanted to ask if it's best to accept the changes directly here in GH or change it locally and push a new commit?

Reply:
I can only tell you my personal preference when coding, which is to make the changes by hand on my local branch, commit locally, then squash the new commit and force-push the branch. That way I keep the PR to a single commit, and can test the patches before committing them (reviewers make mistakes!). However, you are welcome to accept the suggestions via GitHub if you want to try that workflow instead.

Reply:
The local commit / squash / force push approach makes sense.
Will wait for further feedback and testing.

Thanks for taking the time to explain. Cheers.

Reply:
I'd say that GeneFC's testing was pretty thorough, I'd say go ahead and do a final cleanup path and let's merge it.

Reply:
Changed made as suggested.